src/native/managed/cdacreader/Microsoft.Diagnostics.DataContractReader.Contracts/Contracts/ExecutionManager/Helpers/HotColdLookup.cs

```diff
@@ -0,0 +1,134 @@
+
+        // If runtime function is in the cold part, get the associated hot part
+        Debug.Assert((hotIndex & 1) == 1);
+        return _target.Read<uint>(hotColdMap + (ulong)hotIndex * sizeof(uint));
+    }
+
+    public bool TryGetColdFunctionIndex(uint numHotColdMap, TargetPointer hotColdMap, uint runtimeFunctionIndex, out uint functionIndex)
+    {
+        functionIndex = ~0u;
+
+        uint coldIndex;
+        if (!TryLookupHotColdMappingForMethod(numHotColdMap, hotColdMap, runtimeFunctionIndex, out uint _, out coldIndex))
+            return false;
+
+        functionIndex = _target.Read<uint>(hotColdMap + (ulong)coldIndex * sizeof(uint));
+        return true;
+    }
+
+    private bool IsColdCode(uint numHotColdMap, TargetPointer hotColdMap, uint runtimeFunctionIndex)
+    {
+        if (numHotColdMap == 0)
+            return false;
+
+        // Determine if the method index represents a hot or cold part by comparing against the first
+        // cold part index (hot < cold).
+        uint firstColdRuntimeFunctionIndex = _target.Read<uint>(hotColdMap);
+        return runtimeFunctionIndex >= firstColdRuntimeFunctionIndex;
+    }
+
+    // Look up a runtime function index in the hot/cold map. If the function is in the
+    // hot/cold map, return whether the function corresponds to cold code and the hot
+    // and cold lookup indexes for the function.
+    private bool TryLookupHotColdMappingForMethod(
+        uint numHotColdMap,
+        TargetPointer hotColdMap,
+        uint runtimeFunctionIndex,
+        out uint hotIndex,
+        out uint coldIndex)
+    {
+        hotIndex = ~0u;
+        coldIndex = ~0u;
+
+        // HotColdMappingLookupTable::LookupMappingForMethod
+        if (numHotColdMap == 0)
+            return false;
+
+        // Each method is represented by a pair of unsigned 32-bit integers. First is the runtime
+        // function index of the cold part, second is the runtime function index of the hot part.
+        // HotColdMap is these pairs as an array, so the logical size is half the array size.
+        uint start = 0;
+        uint end = (numHotColdMap - 1) / 2;
+
+        bool isColdCode = IsColdCode(numHotColdMap, hotColdMap, runtimeFunctionIndex);
+        int indexCorrection = isColdCode ? 0 : 1;
+
+        // Entries are sorted by the hot part runtime function indices. This also means they are sorted
+        // by the cold part indices, as the cold part is emitted in the same order as hot parts.
+        // Binary search until we get to 10 or fewer items.
+        while (end - start > 10)
+        {
+            uint middle = start + (end - start) / 2;
+            long index = middle * 2 + indexCorrection;
+
+            if (runtimeFunctionIndex < _target.Read<uint>(hotColdMap + (ulong)(index * sizeof(uint))))
+            {
+                end = middle - 1;
+            }
+            else
+            {
+                start = middle;
+            }
+        }
```

Comment:
I'm not convinced this algorithm is needed in the cDAC side. It has utility in many cases, but in this case seems like a premature optimization and not adding a lot of value. Concerns with dropping it for now?